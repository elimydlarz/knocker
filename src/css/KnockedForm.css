textarea {
  display: block;
  width: 100%;
  height: 100px;
}

input[type="text"] {
  width: 100%;
}

input[type="submit"]:disabled {
  color: #D1D1D1;
  border-color: #D1D1D1;
}
