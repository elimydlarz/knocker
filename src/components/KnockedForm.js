import React, { Component } from 'react';
import FlashMessage from './FlashMessage';
import TextInput from './TextInput';
import Submit from './Submit';
import DropDown from './DropDown';
import Login from './Login';
import '../css/KnockedForm.css';

export default class KnockedForm extends Component {
  constructor() {
    super();
    this.state = {
      message: 'Welcome to Knocker!',
      knocked: {},
      password: '',
    };

    this.setPassword = this.setPassword.bind(this);
    this.updateField = this.updateField.bind(this);
    this.updateCheckedOptions = this.updateCheckedOptions.bind(this);
    this.handleSubmitSuccess = this.handleSubmitSuccess.bind(this);
    this.handleSubmitFailure = this.handleSubmitFailure.bind(this);
  }

  setPassword(event) {
    this.setState({ password: event.target.value });
  }

  updateField(event) {
    this.setState({
      knocked: {
        ...this.state.knocked,
        [event.target.name]: event.target.value,
      },
    });
  }

  updateCheckedOptions(event) {
    const add = (list, value) => (list || []).concat(value);
    const remove = (list, value) => list.filter(option => option !== value);

    this.setState({
      knocked: {
        ...this.state.knocked,
        [event.target.name]: event.target.checked
          ? add(this.state.knocked[event.target.name], event.target.value)
          : remove(this.state.knocked[event.target.name], event.target.value)
      }
    })
  }

  handleSubmitSuccess() {
    this.setState({
      message: `Thanks heaps for knocking at ${this.state.knocked.knockeeAddress}.
                You're a great knocker, ${this.state.knocked.knocker}!`,
      knocked: { knocker: this.state.knocked.knocker },
      authorised: true,
    });
  }

  handleSubmitFailure(response) {
    const message = response.status === 401
      ? `We're having trouble authenticating you.
        Make sure you've identified yourself correctly and
        supplied the right password.`
      : 'Whoops! Something unexpected has gone wrong.'
    this.setState({
      message ,
      authorised: false,
    });
  }

  render() {
    return (
      <div className='KnockedForm'>
        <FlashMessage
          message={this.state.message} />
        <form>
          <Login
            user={this.state.knocked.knocker || ''}
            password={this.state.password}
            userHandler={this.updateField}
            passwordHandler={this.setPassword}
            authorised={this.state.authorised}
          />
          <TextInput
            name='knockeeAddress'
            label='Where did you knock?'
            value={this.state.knocked.knockeeAddress || ''}
            changeHandler={this.updateField}
          />
          <TextInput
            name='knockeePerson'
            label='Who did you speak to? (optional)'
            value={this.state.knocked.knockeePerson || ''}
            changeHandler={this.updateField}
          />
          <DropDown
            name='interaction'
            value={this.state.knocked.interaction}
            changeHandler={this.updateField}
            placeholder='What happened?'
            options={[
              'We had a non-meaningful interaction',
              'We had a meaningful interaction',
              'They were busy',
            	'There was a language barrier',
            	'There was no answer',
            	'I received bad information',
            	'The property was inaccessible',
            	'They refused to talk to me',
            ]}
          />
          <DropDown
            name='support'
            value={this.state.knocked.support}
            changeHandler={this.updateField}
            placeholder='What was their level of support? (optional)'
            options={[
            	'Strong',
            	'Weak',
            	'Undecided',
            	'Weakly opposed',
            	'Strongly opposed',
              'Unknown',
            ]}
          />
          <textarea name='issues'
            placeholder={`What issues did they care about? (optional)
e.g. health, jobs, climate change`}
            value={this.state.knocked.issues || ''}
            onChange={this.updateField}
          />
          <textarea name='notes'
            placeholder='Anything else? (optional)'
            value={this.state.knocked.notes || ''}
            onChange={this.updateField}
          />
          <DropDown
            name='followupRequired'
            value={this.state.knocked.followupRequired}
            changeHandler={this.updateField}
            placeholder='Should we have a followup conversation? (optional)'
            options={[
              'Yeah',
              'Nope',
            ]}
          />
          <Submit
            password={this.state.password}
            data={this.state.knocked}
            requiredFields={[
              this.state.knocked.knocker,
              this.state.password,
              this.state.knocked.knockeeAddress,
              this.state.knocked.interaction,
            ]}
            successFn={this.handleSubmitSuccess}
            failureFn={this.handleSubmitFailure}
          />
        </form>
      </div>
    );
  }
}
