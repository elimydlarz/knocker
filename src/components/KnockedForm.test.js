import React from 'react';
import renderer from 'react-test-renderer';
import KnockedForm from './KnockedForm';

describe('KnockedForm', () => {
  it('does not explode', () => {
    const component = renderer.create(<KnockedForm />)
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
