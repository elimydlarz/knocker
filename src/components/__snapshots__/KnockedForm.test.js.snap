exports[`KnockedForm does not explode 1`] = `
<div
  className="KnockedForm">
  <div
    className="FlashMessage">
    Welcome to Knocker!
  </div>
  <form>
    <div
      className="TextInput">
      <label>
        Who knocked?
        <input
          name="knocker"
          onChange={[Function bound updateField]}
          type="text"
          value="" />
      </label>
    </div>
    <div
      className="TextInput">
      <label>
        Where did you knock?
        <input
          name="knockeeAddress"
          onChange={[Function bound updateField]}
          type="text"
          value="" />
      </label>
    </div>
    <div
      className="TextInput">
      <label>
        Who did you speak to?
        <input
          name="knockeePerson"
          onChange={[Function bound updateField]}
          type="text"
          value="" />
      </label>
    </div>
    <div
      className="RadioButtons">
      <label>
        <div
          className="radio-heading">
          What happened?
        </div>
        <label
          className="radio-label">
          <input
            checked={false}
            name="interaction"
            onClick={[Function bound updateField]}
            type="radio"
            value="Non-meaningful interaction" />
          Non-meaningful interaction
        </label>
        <label
          className="radio-label">
          <input
            checked={false}
            name="interaction"
            onClick={[Function bound updateField]}
            type="radio"
            value="Meaningful interaction" />
          Meaningful interaction
        </label>
        <label
          className="radio-label">
          <input
            checked={false}
            name="interaction"
            onClick={[Function bound updateField]}
            type="radio"
            value="Busy" />
          Busy
        </label>
        <label
          className="radio-label">
          <input
            checked={false}
            name="interaction"
            onClick={[Function bound updateField]}
            type="radio"
            value="Language Barrier" />
          Language Barrier
        </label>
        <label
          className="radio-label">
          <input
            checked={false}
            name="interaction"
            onClick={[Function bound updateField]}
            type="radio"
            value="No Answer" />
          No Answer
        </label>
        <label
          className="radio-label">
          <input
            checked={false}
            name="interaction"
            onClick={[Function bound updateField]}
            type="radio"
            value="Bad Info" />
          Bad Info
        </label>
        <label
          className="radio-label">
          <input
            checked={false}
            name="interaction"
            onClick={[Function bound updateField]}
            type="radio"
            value="Inaccessible" />
          Inaccessible
        </label>
        <label
          className="radio-label">
          <input
            checked={false}
            name="interaction"
            onClick={[Function bound updateField]}
            type="radio"
            value="Refused" />
          Refused
        </label>
      </label>
    </div>
    <div
      className="RadioButtons">
      <label>
        <div
          className="radio-heading">
          What was their level of support?
        </div>
        <label
          className="radio-label">
          <input
            checked={false}
            name="support"
            onClick={[Function bound updateField]}
            type="radio"
            value="Unknown" />
          Unknown
        </label>
        <label
          className="radio-label">
          <input
            checked={false}
            name="support"
            onClick={[Function bound updateField]}
            type="radio"
            value="Strong support" />
          Strong support
        </label>
        <label
          className="radio-label">
          <input
            checked={false}
            name="support"
            onClick={[Function bound updateField]}
            type="radio"
            value="Weak support" />
          Weak support
        </label>
        <label
          className="radio-label">
          <input
            checked={false}
            name="support"
            onClick={[Function bound updateField]}
            type="radio"
            value="Undecided" />
          Undecided
        </label>
        <label
          className="radio-label">
          <input
            checked={false}
            name="support"
            onClick={[Function bound updateField]}
            type="radio"
            value="Weak oppose" />
          Weak oppose
        </label>
        <label
          className="radio-label">
          <input
            checked={false}
            name="support"
            onClick={[Function bound updateField]}
            type="radio"
            value="Strong oppose" />
          Strong oppose
        </label>
      </label>
    </div>
    <div
      className="Checkboxes">
      <label>
        <div
          className="checkbox-heading">
          What issues did they care about?
        </div>
        <label
          className="checkbox-label">
          <input
            checked={false}
            name="issues"
            onClick={[Function bound updateCheckedOptions]}
            type="checkbox"
            value="Jobs" />
          Jobs
        </label>
        <label
          className="checkbox-label">
          <input
            checked={false}
            name="issues"
            onClick={[Function bound updateCheckedOptions]}
            type="checkbox"
            value="Cost of living" />
          Cost of living
        </label>
        <label
          className="checkbox-label">
          <input
            checked={false}
            name="issues"
            onClick={[Function bound updateCheckedOptions]}
            type="checkbox"
            value="Economy" />
          Economy
        </label>
        <label
          className="checkbox-label">
          <input
            checked={false}
            name="issues"
            onClick={[Function bound updateCheckedOptions]}
            type="checkbox"
            value="Education " />
          Education 
        </label>
        <label
          className="checkbox-label">
          <input
            checked={false}
            name="issues"
            onClick={[Function bound updateCheckedOptions]}
            type="checkbox"
            value="Health" />
          Health
        </label>
        <label
          className="checkbox-label">
          <input
            checked={false}
            name="issues"
            onClick={[Function bound updateCheckedOptions]}
            type="checkbox"
            value="Climate change" />
          Climate change
        </label>
        <label
          className="checkbox-label">
          <input
            checked={false}
            name="issues"
            onClick={[Function bound updateCheckedOptions]}
            type="checkbox"
            value="Environment" />
          Environment
        </label>
        <label
          className="checkbox-label">
          <input
            checked={false}
            name="issues"
            onClick={[Function bound updateCheckedOptions]}
            type="checkbox"
            value="Refugees" />
          Refugees
        </label>
        <label
          className="checkbox-label">
          <input
            checked={false}
            name="issues"
            onClick={[Function bound updateCheckedOptions]}
            type="checkbox"
            value="Other" />
          Other
        </label>
      </label>
    </div>
    <div
      className="RadioButtons">
      <label>
        <div
          className="radio-heading">
          Should we have a followup conversation?
        </div>
        <label
          className="radio-label">
          <input
            checked={false}
            name="followupRequired"
            onClick={[Function bound updateField]}
            type="radio"
            value="Yeah" />
          Yeah
        </label>
        <label
          className="radio-label">
          <input
            checked={false}
            name="followupRequired"
            onClick={[Function bound updateField]}
            type="radio"
            value="Nope" />
          Nope
        </label>
      </label>
    </div>
    <label>
      Anything else?
      <textarea
        name="notes"
        onChange={[Function bound updateField]}
        value="" />
    </label>
    <div
      className="Submit">
      <input
        disabled={true}
        onClick={[Function bound transmit]}
        type="submit" />
    </div>
  </form>
</div>
`;
